grype/db/curator_test.go

```diff
@@ -447,3 +448,107 @@ func TestCuratorTimeoutBehavior(t *testing.T) {
 		t.Fatalf("timeout exceeded (%v)", failAfter)
 	}
 }
+
+func TestCuratorIsUpdateAvailable(t *testing.T) {
+	testDir := "/tmp/dbDir"
+	listingURL := "http://example.com/metadata/listing.json"
+	tests := []struct {
+		name              string
+		existingMetadata  Metadata
+		listing           Listing
+		forceListingCheck bool
+		wantListingCheck  bool
+		wantUpdate        bool
+		wantErr           bool
+	}{
+		{
+			name:             "skip update check; too soon",
+			existingMetadata: metadataBuiltAt(time.Now().Add(-10 * time.Minute)),
+			wantListingCheck: false,
+			wantUpdate:       false,
+			wantErr:          false,
+		},
+		{
+			name:              "force causes check even if local metadata is very new",
+			existingMetadata:  metadataBuiltAt(time.Now().Add(-10 * time.Minute)),
+			listing:           listingWithEntryBuiltAt(time.Time{}, "http://example.com/payload.tar.gz"),
+			forceListingCheck: true,
+			wantListingCheck:  true,
+			wantUpdate:        false,
+			wantErr:           false,
+		},
+		{
+			name:             "check for update but not needed",
+			existingMetadata: metadataBuiltAt(time.Now().Add(-50 * time.Hour)), // old listing file
+			listing:          listingWithEntryBuiltAt(time.Time{}, "http://example.com/payload.tar.gz"),
+			wantListingCheck: true,
+			wantUpdate:       false,
+			wantErr:          false,
+		},
+		{
+			name:             "update needed",
+			wantListingCheck: true,
+			wantUpdate:       true,
+			wantErr:          false,
+			existingMetadata: metadataBuiltAt(time.Now().Add(-50 * time.Hour)), // old listing file
+			listing:          listingWithEntryBuiltAt(time.Now().Add(-1*time.Hour), "http://example.com/payload.tar.gz"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			files := map[string]string{}
+			b, err := json.Marshal(tt.listing)
+			require.NoError(t, err)
+			files[listingURL] = string(b)
+			dirs := map[string]string{}
+			fs := afero.NewMemMapFs()
+			fs.MkdirAll(path.Join(testDir, "5"), 0755)
+			f, err := fs.Create(path.Join(testDir, "5", "metadata.json"))
+			require.NoError(t, err)
+			defer f.Close()
+			err = json.NewEncoder(f).Encode(tt.existingMetadata)
+			require.NoError(t, err)
+			getter := newTestGetter(fs, files, dirs)
```

Comment:
I had some issues getting this working in all the scenarios I wanted to test because it seemed as though I couldn't mock every call and the getter code was ultimately calling `os.Open`, instead of using the `afero.FS`. If you happen to run into similar issues, I [wrote a mocked but functional server to emulate the listing file / db download](https://github.com/anchore/grype/pull/1247/files#diff-bc68f6371e1fd9156df23d07f9939a6bf90495b4d41668489f65383700cf710cR383).

Reply:
I agree testing with a local http server is a cleaner solution. I'll wait for that PR to get in and then rebase and change these tests to use that, assuming we still decide we want this behavior.